ibis/pandas/trace.py

```diff
@@ -0,0 +1,163 @@
+DEBUG:ibis.pandas.trace:           execute_until_in_scope TableColumn
+DEBUG:ibis.pandas.trace:             execute_until_in_scope PandasTable
+DEBUG:ibis.pandas.trace:             execute_until_in_scope PandasTable 0:00:00.000061
+DEBUG:ibis.pandas.trace:             execute_table_column_df_or_df_groupby TableColumn
+DEBUG:ibis.pandas.trace:             execute_table_column_df_or_df_groupby TableColumn 0:00:00.000304  # noqa: E501
+DEBUG:ibis.pandas.trace:           execute_until_in_scope TableColumn 0:00:00.000584
+DEBUG:ibis.pandas.trace:           execute_udf_node ElementWiseVectorizedUDF
+DEBUG:ibis.pandas.trace:           execute_udf_node ElementWiseVectorizedUDF 0:00:05.019173
+DEBUG:ibis.pandas.trace:         execute_until_in_scope ElementWiseVectorizedUDF 0:00:05.052604  # noqa: E501
+DEBUG:ibis.pandas.trace:       main_execute ElementWiseVectorizedUDF 0:00:05.052819
+DEBUG:ibis.pandas.trace:     execute_selection_dataframe Selection 0:00:05.054894
+DEBUG:ibis.pandas.trace:   execute_until_in_scope Selection 0:00:05.055662
+DEBUG:ibis.pandas.trace: main_execute Selection 0:00:05.056556
+
+"""
+
+_logger = logging.getLogger('ibis.pandas.trace')
+
+# A list of funcs that is traced
+_trace_funcs = set()
+_trace_root = "main_execute"
+_TRACE_CONFIG = 'pandas.enable_trace'
+
+
+def enable():
+    """Enable tracing."""
+    set_option(_TRACE_CONFIG, True)
+    logging.getLogger('ibis.pandas.trace').setLevel(logging.DEBUG)
+
+
+def _log_trace(func, start=None):
+    level = 0
+    current_frame = None
+
+    # Increase the current level for each traced function in the stackframe
+    # This way we can visualize the call stack.
+    for frame, _ in traceback.walk_stack(None):
+        current_frame = current_frame if current_frame is not None else frame
+        func_name = frame.f_code.co_name
+        if func_name in _trace_funcs:
+            level += 1
+
+    # We can assume we have 'args' because we only call _log_trace inside
+    # trace or TraceDispatcher.resgister
+    current_op = current_frame.f_locals['args'][0]
+
+    # If the first argument is a Expr, we print its op because it's more
+    # informative.
+    if isinstance(current_op, ir.Expr):
+        current_op = current_op.op()
+
+    _logger.debug(
+        f"{'  ' * level} {func.__name__} {type(current_op).__qualname__} "
+        f"{datetime.now() - start if start else ''}"
+    )
+
+
+def trace(func):
+    """ Return a function decorator that wraped the decorated function with
+    tracing.
+    """
+    _trace_funcs.add(func.__name__)
+
+    @functools.wraps(func)
+    def traced_func(*args, **kwargs):
+        trace_enabled = get_option(_TRACE_CONFIG)
+
+        if not trace_enabled:
+            return func(*args, **kwargs)
+        else:
+            start = datetime.now()
+            _log_trace(func)
+            res = func(*args, **kwargs)
+            _log_trace(func, start)
+            return res
+
+    return traced_func
+
+
+class TraceDispatcher(Dispatcher):
```

Comment:
why do we need this additional complexity here? can you just modify Dispatcher to add (if the config is on)? isn't that the point of using the option.

Reply:
Dispatcher is from multipledispatch library..